helix-core/src/selection.rs

```diff
@@ -234,8 +235,8 @@ impl Selection {
                     let from = prev.from();
                     let to = std::cmp::max(range.to(), prev.to());
 
-                    if i <= primary_index {
-                        primary_index -= 1
+                    if range == primary {
+                        primary_index = result_len - 1;
```

Comment:
I don't think this fix is correct. It will always set the primary index to `len() - 1`. Whereas what we wanted to do here is move the primary range back by 1 slot if a range is merged before.

It's a direct port of https://github.com/codemirror/state/blob/d48e95808afc8bab6a61cf353a8892405fe85033/src/selection.ts#L204

Reply:
The idea in this change is to set the primary index to be what the merged range will be. So that last item in result. But I'm not 100% confident the idea is correct, since I don't understand the ranges here correctly, so you might be right.

What I'm still quite sure is that the old code *is* incorrect. It will compare `primary_index` to `i`, but they are not really related. `i` is the one that is being looped in the original range, but the new range might already be much smaller, if ranges have been merged already.

Reply:
It checks if the primary index is further ahead than `i`.

Imagine we have 4 ranges, 0-3, and the primary index is 2. 0 and 1 overlap, so they get merged into a single range. That means the primary selection will now sit at index 1.

Reply:
My thinking is that the issue could occur as follows.

We have 4 ranges 0-3, primary index is 2. 0 and 1 overlap, primary index is decremented to 1. Then that new merged range overlaps with original 2. `i` is now 2, while `primary_index` is still 1 and won't get decremented. Then also that last range overlaps and gets merged. `primary_index` again won't decrement. Now we have 1 range and `primary_index` of 1, so index is out of bounds. Not sure if something like this could happen (since I'm not selection expert here), but the crash deifinitely happened somehow. Maybe something similar to this description.

Anyway, my idea wasn't to set the selection to the *last* index, but to the merged one. Not that primary_index is only assigned if merged range is primary *and* we are actually merging it to the previous. Then we set the primary index to be the previous item (current last in results).

And yeah, I'm not happy with the resulting code, but this was the one that made somewhat sense to me that I could do in reasonable amount of time.

Reply:
Oh I see, so the `i` in the comparison needs to be calculated based on the position in the new result set.

@cessen looks like you fixed this in your refactor PR? https://github.com/helix-editor/helix/pull/376/files#diff-f218cebd639a867087e2a7abc84160d45888ad5bcff4ab10fcf7217cb16586e9R320-R333

I was able to reproduce this with:

```rust
        // it correctly calculates a new primary index
        let sel = Selection::new(
            smallvec![Range::new(0, 2), Range::new(1, 5), Range::new(4, 7)],
            2,
        );

        let res = sel
            .ranges
            .into_iter()
            .map(|range| format!("{}/{}", range.anchor, range.head))
            .collect::<Vec<String>>()
            .join(",");

        assert_eq!(res, "0/7");
        assert_eq!(sel.primary_index, 0); // code calculates 1

```

Reply:
Yeah, I did a big clean up of `Range` and `Selection` as the first steps in my refactor, and that included fixing this.

I don't think it hurts to merge this, though.  It's probably still going to be a bit before we can merge my PR, and it'll be nice to have the fix in the mean time.